open_spiel/games/checkers.h

```diff
@@ -0,0 +1,134 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPEN_SPIEL_GAMES_CHECKERS_H_
+#define OPEN_SPIEL_GAMES_CHECKERS_H_
+
+// Implementation of the board game Checkers.
+// https://en.wikipedia.org/wiki/Checkers
+//
+// Some notes about this implementation:
+// - Drawing: 
+//     Game is drawn if no pieces have been removed in 40 moves
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "open_spiel/spiel.h"
+
+namespace open_spiel {
+namespace checkers {
+
+inline constexpr int kNumPlayers = 2;
+inline constexpr int kDefaultRows = 8;
+inline constexpr int kDefaultColumns = 8;
+inline constexpr int kMaxMovesWithoutCapture = 40;
+inline constexpr int kCellStates = 5;  // Empty, White, WhiteCrowned, Black and BlackCrowned.
+
+// State of a cell.
+enum class CellState {
+  kEmpty,         // Represented by ' '.
+  kWhite,         // Represented by 'o'.
+  kBlack,         // Represented by '+'.
+  kWhiteCrowned,  // Represented by '8'.
+  kBlackCrowned,  // Represented by '*'.
+};
+
+// Types of moves.
+enum MoveType {
+  kNormal = 0,
+  kCapture = 1,
+};
+
+// Types of pieces.
+enum PieceType {
+  kMan = 0,
+  kKing = 1,
+};
+
+// State of an in-play game.
+class CheckersState : public State {
+ public:
+  explicit CheckersState(std::shared_ptr<const Game> game, int rows,
+                        int columns);
+  Player CurrentPlayer() const override {
+    return IsTerminal() ? kTerminalPlayerId : current_player_;
+  }
+  std::string ActionToString(Player player, Action action_id) const override;
+  std::string ToString() const override;
+  bool IsTerminal() const override;
+  std::vector<double> Returns() const override;
+  std::string InformationStateString(Player player) const override;
+  std::string ObservationString(Player player) const override;
+  void ObservationTensor(Player player,
+                         absl::Span<float> values) const override;
+  std::unique_ptr<State> Clone() const override {
+    return std::unique_ptr<State>(new CheckersState(*this));
+  }
+  void UndoAction(Player player, Action action) override;
+  bool InBounds(int row, int column) const;
+  void SetCustomBoard(const std::string board_string);
+  void SetBoard(int row, int column, CellState state) {    
+    board_[row * columns_ + column] = state;
+  }
+  CellState BoardAt(int row, int column) const {
+    return board_[row * columns_ + column];
+  }
+  std::vector<Action> LegalActions() const override;
+
+ protected:
+  void DoApplyAction(Action action) override;
+
+ private:
+  int ObservationPlane(CellState state, Player player) const;
+
+  Player current_player_ = 0;  // Player zero (White, 'o') goes first.
+  Player outcome_ = kInvalidPlayer;
+  int multiple_jump_piece_ = 0; // Piece in the board who can do multiple jump. Represented by row * rows_ + column
```

Comment:
Can you say a bit more in the comments here about how this variable is used or why it's needed?

Reply:
This variable is used to indicate that the last moved piece can make an additional jump (multiple jumps). In that case LegalActions is filtered to only represent moves available for this piece.

I'm not entirely happy with the implementation since I had to use a global variable, if you have other thoughts, do share them

Reply:
Ok great. But, what I meant was: could you elaborate on that in the comment?

Also, can we please clarify what version of the rules you are using (at the top of the header file)? I wasn't aware of forced multi-captures, e.g. I don't see anything about it mentioned on https://en.wikipedia.org/wiki/Checkers. Maybe this is a common convention used in computer Checkers? If so, please clarify that in the documentation at the top of the file.

Reply:
It's mentioned in the Wiki link that
> When capturing an opponent's piece is possible, capturing is mandatory in most official rules.
That's why I added those. About the 40 move rule, I could find some mentions of it online
> http://www.flyordie.com/games/help/checkers/en/games_rules_checkers.html